### PR TITLE
Fix hybrid query loading of expired entities (#13194)

### LINKED ISSUE #13194: Error querying indexed cache with lifespan after expiration ('instance cannot be null')
### Describe the bug

Some type of queries (likely aggregation) fail when they hit recently expired entries with lifespan. The error reported is:
java.lang.IllegalArgumentException: instance cannot be null.
An example of such queries is  'select count(status)  from test.Task t group by status'

### Expected behavior

_No response_

### Actual behavior

_No response_

### Steps to reproduce with Infinispan 15.0.10

From the console:

1. Create schema test.proto (see schema below)
2. Create the cache TEST_TASKS (see xml below)
3. Add one entry to the cache, with **LIFESPAN**=30 secs
```
{
  "_type": "test.Task",
  "id": 111,
  "type": "test 1",
  "status": "status1"
}
```

4. Run the following query multiple times until it fails during expiration:
   select count(status)  from test.Task t group by status
   
5. When the entry expires, a REST error is generated:
   {"error":{"message":"Error executing search","cause":"instance cannot be null"}}

Unfortunately there's no error thrown by Infinispan (likely the REST endpoint is swallowing the exception).

This also happens from Hot Rod (see exception details in additional information)

### Protobuf schema
```
// test.proto


syntax = "proto2";
package test;

/**
 * @Indexed
 */
message Task {

   /**
    * @Basic(projectable=true, sortable=true)
    */
   required int64 id = 1;

   required string type = 2;

   /**
    * @Keyword
    */
   optional string status = 5 [default = "WAITING"];

}

```

### Cache configuration
```
<?xml version="1.0"?>
<distributed-cache name="TEST_TASKS" owners="1" mode="SYNC" remote-timeout="17500" statistics="true">
	<encoding>
		<key media-type="application/x-protostream"/>
		<value media-type="application/x-protostream"/>
	</encoding>
	<locking concurrency-level="1000" acquire-timeout="15000" striping="false"/>
	<expiration lifespan="-1"/>
	<memory storage="HEAP" max-count="1000"/>
	<persistence passivation="false">
		<file-store purge="false" read-only="false"/>
	</persistence>
	<indexing enabled="true" storage="filesystem">
		<indexed-entities>
			<indexed-entity>test.Task</indexed-entity>
		</indexed-entities>
	</indexing>
	<state-transfer timeout="60000"/>
</distributed-cache>
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

2024-11-02 16:30:03,714 ERROR [o.i.s.h.BaseRequestProcessor] ISPN005003: Exception reported java.lang.IllegalArgumentException: instance cannot be null
at org.infinispan.objectfilter.impl.ObjectFilterImpl.filter(ObjectFilterImpl.java:199)
at org.infinispan.query.core.impl.MetadataHybridQuery.filter(MetadataHybridQuery.java:48)
at org.infinispan.query.core.impl.MappingEntryIterator.transform(MappingEntryIterator.java:63)
at org.infinispan.query.core.impl.MappingEntryIterator.updateNext(MappingEntryIterator.java:45)
at org.infinispan.query.core.impl.MappingEntryIterator.hasNext(MappingEntryIterator.java:27)
at org.infinispan.query.core.impl.MappingIterator.updateNext(MappingIterator.java:54)
at org.infinispan.query.core.impl.MappingIterator.hasNext(MappingIterator.java:38)
at java.base/java.util.Iterator.forEachRemaining(Iterator.java:132)
at org.infinispan.query.core.impl.AggregatingQuery.getBaseIterator(AggregatingQuery.java:59)
at org.infinispan.query.core.impl.HybridQuery.getInternalIterator(HybridQuery.java:57)
at org.infinispan.query.core.impl.BaseEmbeddedQuery.executeInternal(BaseEmbeddedQuery.java:101)
at org.infinispan.query.core.impl.BaseEmbeddedQuery.execute(BaseEmbeddedQuery.java:80)
at org.infinispan.query.core.impl.HybridQuery.execute(HybridQuery.java:68)
at org.infinispan.query.core.impl.DelegatingQuery.execute(DelegatingQuery.java:110)
at org.infinispan.query.remote.impl.BaseRemoteQueryManager.executeQuery(BaseRemoteQueryManager.java:63)
at org.infinispan.query.remote.impl.QueryFacadeImpl.query(QueryFacadeImpl.java:45)
at org.infinispan.server.hotrod.HotRodServer.query(HotRodServer.java:151)
at org.infinispan.server.hotrod.CacheRequestProcessor.queryInternal(CacheRequestProcessor.java:542)
at org.infinispan.server.hotrod.CacheRequestProcessor.lambda$query$21(CacheRequestProcessor.java:537)
at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
at java.base/java.lang.Thread.run(Thread.java:1583)

### PR DESCRIPTION
**Pull request by fax4ever**
closes #13194

### CODE REVIEW

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+      config.encoding().value().mediaType(MediaType.APPLICATION_PROTOSTREAM);
+      config.locking().concurrencyLevel(1000).lockAcquisitionTimeout(15000);
+      config.expiration().lifespan(-1);
+      config.memory().storage(StorageType.HEAP).maxCount(1000);
+      config.persistence().passivation(false)
+            .addStore(new DummyInMemoryStoreConfigurationBuilder(config.persistence()));
+      config.indexing().enable()
+            .storage(FILESYSTEM).path(indexDirectory)
+            .addIndexedEntity("model.Task");
+      config.clustering().stateTransfer().timeout(60000);
```
Comment (pruivo) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java:76]: can we remove any unnecessary configuration? Most of them seem irrelevant and you only need indexing.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import org.infinispan.configuration.cache.StorageType;
+import org.infinispan.persistence.dummy.DummyInMemoryStoreConfigurationBuilder;
+import org.infinispan.protostream.SerializationContextInitializer;
+import org.infinispan.query.model.Task;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.Test;
+
+@Test(groups = "functional", testName = "client.hotrod.query.aggregation.RemoteClusteredAggregationExpiringTest")
+@TestForIssue(githubKey = "13194")
+public class RemoteClusteredAggregationExpiringTest extends MultiHotRodServersTest {
```
Comment (pruivo) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java:30]: Can it be reproduced in a non-clustered environment?

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+   }
+
+   @Test
+   public void test() throws Exception {
+      RemoteCache<Integer, Task> remoteCache = clients.get(0).getCache();
+      for (int i=0; i<12; i++) {
+         int module = i % 4;
+         remoteCache.put(i, new Task(100 + i, "type-" + module, "status-" + module, "label-" + module),
+               1, TimeUnit.SECONDS);
+         Thread.sleep(10);
```
Comment (pruivo) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java:107]: why sleeping?

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+         //delete the index otherwise it will mess up the index for next tests
+         Util.recursiveFileRemove(indexDirectory);
+      }
+      super.destroy();
+   }
+
+   @Test
+   public void test() throws Exception {
+      RemoteCache<Integer, Task> remoteCache = clients.get(0).getCache();
+      for (int i=0; i<12; i++) {
```
Comment (pruivo) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java:103]: Isn't 2 entries enough? one immortal and another expired and check if the query returns only the immortal entry.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+         int module = i % 4;
+         remoteCache.put(i, new Task(100 + i, "type-" + module, "status-" + module, "label-" + module),
+               1, TimeUnit.SECONDS);
+         Thread.sleep(10);
+      }
+
+      int continueToRunFor = 3;
+      while (continueToRunFor > 0) {
```
Comment (pruivo) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java:111]: why looping?

File: commons-test/src/main/java/org/infinispan/commons/test/annotation/TestForIssue.java [hunk 6]
```diff
@@ -21,6 +21,13 @@
...
     */
-   String[] jiraKey();
+   String[] jiraKey() default "";
+
+   /**
+    * The keys of the GitHub issues tested.
+    *
+    * @return The GitHub issue keys
+    */
+   String[] githubKey() default "";
```
Comment (pruivo) [inline commons-test/src/main/java/org/infinispan/commons/test/annotation/TestForIssue.java:31]: It looks like you're the only one using this annotation :smile:

File: query-core/src/main/java/org/infinispan/query/core/impl/MetadataHybridQuery.java [hunk 7]
```diff
@@ -54,7 +54,15 @@ protected ClosableIteratorWithCount<ObjectFilter.FilterResult> getInternalIterat
    }
 
    private ObjectFilter.FilterResult filter(EntityEntry<Object, S> entry) {
-      ObjectFilter.FilterResult filter = objectFilter.filter(entry.key(), entry.value(), entry.metadata());
+      S value = entry.value();
```
Comment (pruivo) [inline query-core/src/main/java/org/infinispan/query/core/impl/MetadataHybridQuery.java:57]: Don't you need to fix `HybridQuery` too? Or does it make sense to fix this in the `ObjectFilter` by rejecting anything with `value == null`?

**Review by pruivo — CHANGES_REQUESTED** [review]

File: commons-test/src/main/java/org/infinispan/commons/test/annotation/TestForIssue.java (hunk 6 above)
Comment (wburns) [reply commons-test/src/main/java/org/infinispan/commons/test/annotation/TestForIssue.java:31]: I never even knew this existed! Also I guess it needs to be updated as we aren't using JIRA anymore.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java (hunk 1 above)
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java:76]: I added them to copy the reproducer provided with the issue. But, sure, we can simplify

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java (hunk 2 above)
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java:30]: If I remember correcty, it is not possible. But I can check

File: query-core/src/main/java/org/infinispan/query/core/impl/MetadataHybridQuery.java (hunk 7 above)
Comment (wburns) [reply query-core/src/main/java/org/infinispan/query/core/impl/MetadataHybridQuery.java:57]: Yeah, it isn't quite clear which cases fetch the Cache and when others don't.

Been discussing with @fax4ever directly, but I also worry that cases when queries do access the Cache they would be refreshing last access times for things like eviction and expiration, which may or may not be desirable. From a consistency standpoint I would think is undesirable.

Comment (pruivo) [reply query-core/src/main/java/org/infinispan/query/core/impl/MetadataHybridQuery.java:57]: Glad I'm not the only one confused :shushing_face: 

hmm... I think it is ok to update the access time (eviction and expiration). I bet some projects use only the query API and not `cache.get()` :smiley: 


We may introduce flags to control the behavior (if users ask for it)
* Update the access time (expiration)
* Update the eviction position
* Keep the entry in memory if it is loaded from the cache-loader

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java (hunk 3 above)
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java:107]: to spread the expiration over the time

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java (hunk 2 above)
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java:30]: I checked the issue is not reproducible without the clustering configuration.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java (hunk 4 above)
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java:103]: I think that if you want to cover a little the case 12 is a number that is quite reasonable.
We want to test more interactions and more expirations.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java (hunk 5 above)
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteClusteredAggregationExpiringTest.java:111]: we want to do 3 more queries after the variation to be sure that everything goes well in the further itractions as well

File: commons-test/src/main/java/org/infinispan/commons/test/annotation/TestForIssue.java (hunk 6 above)
Comment (fax4ever) [reply commons-test/src/main/java/org/infinispan/commons/test/annotation/TestForIssue.java:31]: Yes I stealed it from the Hibernate team. The problem is that we don't have GitHub references for old issues.

File: query-core/src/main/java/org/infinispan/query/core/impl/MetadataHybridQuery.java (hunk 7 above)
Comment (fax4ever) [reply query-core/src/main/java/org/infinispan/query/core/impl/MetadataHybridQuery.java:57]: the validity (consistency) we can guarantee during the asynchronous moment in which the index still has a reference for an expired entity is limited (also in the literature!).
I think that for the answer we return during the asynchronous moment could be acceptable to count or not all the contribution.

**Comment by fax4ever** [comment]
Thank you @pruivo and @wburns for the review

**Comment by pruivo** [comment]
@fax4ever this is my suggestion for the test without sleeping/loops:
https://github.com/pruivo/infinispan/commit/4e67b750f00d721f7cd07cc4de18c8a501f894cd

**Comment by fax4ever** [comment]
@pruivo I don't agree with the change you're proposing. Sorry. I still believe that reducing from 12 to 2 insert would be wrong and also about the rest.

**Comment by fax4ever** [comment]
Anyway using a time service it is a good idea, so I'm going to steal it. :P

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
merged! thanks @fax4ever !

**Comment by karesti** [comment]
@alanfx

**Pull request merged**